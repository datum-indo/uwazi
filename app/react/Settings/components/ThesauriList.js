/** @format */

import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { I18NLink, t } from 'app/I18N';
import {
  checkThesaurusCanBeDeleted,
  checkThesaurusCanBeClassified,
  deleteThesaurus,
  disableClassification,
  enableClassification,
} from 'app/Thesauri/actions/thesaurisActions';
import { Icon } from 'UI';

import sortThesauri from '../utils/sortThesauri';

export class ThesauriList extends Component {
  getThesaurusSuggestionActions(thesaurus) {
    if (!thesaurus.enable_classification && thesaurus.model_available) {
      return (
        <button
          onClick={this.enableClassification.bind(this, thesaurus)}
          className="btn btn-success btn-xs"
          type="button"
        >
          <Icon icon="toggle-on" />
          &nbsp;
          <span>{t('System', 'Enable')}</span>
        </button>
      );
    }
    if (thesaurus.enable_classification) {
      const view = (
        <div className="thesauri-list vertical-line">
          <span className="thesaurus-suggestion-count">
            {thesaurus.suggestions ? thesaurus.suggestions : 'No'}&nbsp;
            {t('System', 'documents to be reviewed')}
          </span>
          <I18NLink
            to={`/settings/dictionaries/cockpit/${thesaurus._id}`}
            className="btn btn-primary btn-xs"
          >
            <Icon icon="search" />
            &nbsp;
            <span>{t('System', 'View Suggestions')}</span>
          </I18NLink>
        </div>
      );
      return view;
    }
    return null;
  }

  getThesaurusModifyActions(thesaurus) {
    return (
      <div>
        <I18NLink
          to={`/settings/dictionaries/edit/${thesaurus._id}`}
          className="btn btn-default btn-xs"
        >
          <Icon icon="pencil-alt" />
          &nbsp;
          <span>{t('System', 'Edit')}</span>
        </I18NLink>
        {'  '}
        <button
          onClick={this.deleteThesaurus.bind(this, thesaurus)}
          className="btn btn-danger btn-xs template-remove"
          type="button"
        >
          <Icon icon="trash-alt" />
          &nbsp;
          <span>{t('System', 'Delete')}</span>
        </button>
      </div>
    );
  }

  deleteThesaurus(thesaurus) {
    return this.props
      .checkThesaurusCanBeDeleted(thesaurus)
      .then(() => {
        this.context.confirm({
          accept: () => {
            this.props.deleteThesaurus(thesaurus);
          },
          title: `Confirm delete thesaurus: ${thesaurus.name}`,
          message: 'Are you sure you want to delete this thesaurus?',
        });
      })
      .catch(() => {
        this.context.confirm({
          accept: () => {},
          noCancel: true,
          title: `Cannot delete thesaurus: ${thesaurus.name}`,
          message: 'This thesaurus is being used in document types and cannot be deleted.',
        });
      });
  }

  async disableClassification(thesaurus) {
    this.props.disableClassification(thesaurus).catch(() => {
      this.context.confirm({
        accept: () => {},
        noCancel: true,
        title: `Cannot disable classification for thesaurus: ${thesaurus.name}`,
        message: 'Unable to disable classification.',
      });
    });
  }

  async enableClassification(thesaurus) {
    try {
      const canBeEnabled = await this.props.checkThesaurusCanBeClassified(thesaurus);
      if (canBeEnabled) {
        this.props.enableClassification(thesaurus);
      }
    } catch (error) {
      this.context.confirm({
        accept: () => {},
        noCancel: true,
        title: `Cannot enable classification for thesaurus: ${thesaurus.name}`,
        message: 'This thesaurus does not have its topic classification models in a good state.',
      });
    }
  }

  thesaurusNode(thesaurus) {
    return (
      <tr key={thesaurus.name}>
        <th scope="row">
          <I18NLink to={`/settings/dictionaries/edit/${thesaurus._id}`}>{thesaurus.name}</I18NLink>
        </th>
        <td>{this.getThesaurusSuggestionActions(thesaurus)}</td>
        <td>{this.getThesaurusModifyActions(thesaurus)}</td>
      </tr>
    );
  }

  render() {
    return (
      <div className="panel panel-default">
        <div className="panel-heading">{t('System', 'Thesauri')}</div>
        <div className="thesauri-list">
          <table>
            <thead>
              <tr>
                <th className="nameCol" scope="col" />
                <th scope="col" />
                <th scope="col" />
              </tr>
            </thead>
            <tbody>
              {sortThesauri(this.props.dictionaries.toJS()).map(thesaurus =>
                this.thesaurusNode(thesaurus)
              )}
            </tbody>
          </table>
        </div>
        <div className="settings-footer">
          <I18NLink to="/settings/dictionaries/new" className="btn btn-success">
            <Icon icon="plus" />
            <span className="btn-label">{t('System', 'Add thesaurus')}</span>
          </I18NLink>
        </div>
      </div>
    );
  }
}

ThesauriList.propTypes = {
  dictionaries: PropTypes.object,
  deleteThesaurus: PropTypes.func.isRequired,
  disableClassification: PropTypes.func.isRequired,
  enableClassification: PropTypes.func.isRequired,
  checkThesaurusCanBeClassified: PropTypes.func.isRequired,
  checkThesaurusCanBeDeleted: PropTypes.func.isRequired,
};

ThesauriList.contextTypes = {
  confirm: PropTypes.func,
};

export function mapStateToProps(state) {
  return { dictionaries: state.dictionaries };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      deleteThesaurus,
      checkThesaurusCanBeDeleted,
      disableClassification,
      enableClassification,
      checkThesaurusCanBeClassified,
    },
    dispatch
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(ThesauriList);
