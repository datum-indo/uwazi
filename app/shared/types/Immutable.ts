/** @format */

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? {
      toJS(): Elem[];
      get(i: number): IImmutable<Elem>;
      findIndex(fn: (e: IImmutable<Elem>) => boolean): number;
      filter(fn: (e: IImmutable<Elem>) => boolean): IImmutable<T>;
      find(fn: (e: IImmutable<Elem>) => boolean): IImmutable<Elem> | undefined;
      map<T2>(fn: (e: IImmutable<Elem>) => T2): IImmutable<T2[]>;
      reduce<R>(fn: (r: R, e: IImmutable<Elem>) => R, s: R): R;
    }
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
