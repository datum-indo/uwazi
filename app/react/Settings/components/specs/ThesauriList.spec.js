/** @format */
import { shallow } from 'enzyme';
import Immutable from 'immutable';
import React from 'react';

import { ThesauriList } from '../ThesauriList';

describe('ThesaurisList', () => {
  let component;
  let props;
  let context;

  beforeEach(() => {
    props = {
      deleteThesaurus: jasmine.createSpy('deleteThesaurus').and.returnValue(Promise.resolve()),
      disableClassification: jasmine
        .createSpy('disableClassification')
        .and.returnValue(Promise.resolve()),
      enableClassification: jasmine
        .createSpy('enableClassification')
        .and.returnValue(Promise.resolve()),
      checkThesaurusCanBeClassified: jasmine
        .createSpy('checkThesaurusCanBeClassified')
        .and.returnValue(Promise.resolve()),
      checkThesaurusCanBeDeleted: jasmine
        .createSpy('checkThesaurusCanBeDeleted')
        .and.returnValue(Promise.resolve()),
      dictionaries: Immutable.fromJS([
        {
          _id: 'thesaurusUnderscoreId1',
          name: 'Continents',
          model_available: false,
          values: [
            {
              _id: 'valueUnderscoreId1',
              label: 'Africa',
              id: 'valueId1',
            },
            {
              _id: 'valueUnderscoreId2',
              label: 'Asia',
              id: 'valueId2',
            },
            {
              _id: 'valueUnderscoreId3',
              label: 'North America',
              id: 'valueId3',
            },
            {
              _id: 'valueUnderscoreId4',
              label: 'South America',
              id: 'valueId4',
            },
            {
              _id: 'valueUnderscoreId5',
              label: 'Oceania',
              id: 'valueId5',
            },
            {
              _id: 'valueUnderscoreId6',
              label: 'Europe',
              id: 'valueId6',
            },
          ],
        },
        {
          _id: 'thesaurusUnderscoreId2',
          name: 'Issues',
          model_available: true,
          enable_classification: true,
          suggestions: 3,
          values: [
            {
              _id: 'valueUnderscoreId1',
              label: 'Detention',
              id: 'valueId1',
            },
            {
              _id: 'valueUnderscoreId2',
              label: 'Elections',
              id: 'valueId2',
            },
            {
              _id: 'valueUnderscoreId3',
              label: 'Freedom of Movement',
              id: 'valueId3',
            },
          ],
        },
      ]),
    };

    context = {
      confirm: jasmine.createSpy('confirm'),
    };
  });

  const render = () => {
    component = shallow(<ThesauriList {...props} />, { context });
  };

  describe('render', () => {
    it('should match the snapshot', () => {
      render();
      expect(component).toMatchSnapshot();
    });

    it('should render suggestions nodes when suggestions exist', () => {
      render();
      const renderedContexts = component.find('table');
      expect(renderedContexts.length).toBe(1);
      expect(renderedContexts.find({ scope: 'row' }).length).toBe(2);
      expect(renderedContexts.find('td').find('.vertical-line').length).toBe(1);
      expect(
        renderedContexts
          .find('td')
          .find('.vertical-line')
          .containsMatchingElement(<span>View suggestions</span>)
      ).toBeTruthy();
      expect(
        renderedContexts
          .find('.thesaurus-suggestion-count')
          .contains(
            <span className="thesaurus-suggestion-count">{3}&nbsp;documents to be reviewed</span>
          )
      ).toBeTruthy();
    });
  });

  describe('classification', () => {
    it('should validate the classifier model before enabling', done => {
      render();
      component
        .instance()
        .enableClassification({ _id: 'thesaurusUnderscoreId2', name: 'Issues' })
        .then(() => {
          expect(props.checkThesaurusCanBeClassified).toHaveBeenCalled();
          done();
        });
      component.instance().enableClassification({});
    });

    it('should confirm  before deleting the thesaurus', done => {
      render();
      component
        .instance()
        .deleteThesaurus({ _id: 'thesaurusUnderscoreId2', name: 'Issues' })
        .then(() => {
          expect(context.confirm).toHaveBeenCalled();
          expect(props.checkThesaurusCanBeDeleted).toHaveBeenCalled();
          done();
        });
      component.instance().enableClassification({});
    });
  });
});
