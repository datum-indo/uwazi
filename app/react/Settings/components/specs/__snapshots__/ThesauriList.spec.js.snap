// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThesaurisList render should match the snapshot 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    Thesauri
  </div>
  <div
    className="thesauri-list"
  >
    <table>
      <thead>
        <tr>
          <th
            className="nameCol"
            scope="col"
          />
          <th
            scope="col"
          />
          <th
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr
          key="Continents"
        >
          <th
            scope="row"
          >
            <Connect(I18NLink)
              to="/settings/dictionaries/edit/thesaurusUnderscoreId1"
            >
              Continents
            </Connect(I18NLink)>
          </th>
          <td />
          <td>
            <div>
              <Connect(I18NLink)
                className="btn btn-default btn-xs"
                to="/settings/dictionaries/edit/thesaurusUnderscoreId1"
              >
                <Connect(Icon)
                  icon="pencil-alt"
                />
                 
                <span>
                  Edit
                </span>
              </Connect(I18NLink)>
                
              <button
                className="btn btn-danger btn-xs template-remove"
                onClick={[Function]}
                type="button"
              >
                <Connect(Icon)
                  icon="trash-alt"
                />
                 
                <span>
                  Delete
                </span>
              </button>
            </div>
          </td>
        </tr>
        <tr
          key="Issues"
        >
          <th
            scope="row"
          >
            <Connect(I18NLink)
              to="/settings/dictionaries/edit/thesaurusUnderscoreId2"
            >
              Issues
            </Connect(I18NLink)>
          </th>
          <td>
            <div
              className="thesauri-list vertical-line"
            >
              <span
                className="thesaurus-suggestion-count"
              >
                3
                 
                documents to be reviewed
              </span>
              <Connect(I18NLink)
                className="btn btn-primary btn-xs"
                to="/settings/dictionaries/cockpit/thesaurusUnderscoreId2"
              >
                <Connect(Icon)
                  icon="search"
                />
                 
                <span>
                  View Suggestions
                </span>
              </Connect(I18NLink)>
            </div>
          </td>
          <td>
            <div>
              <Connect(I18NLink)
                className="btn btn-default btn-xs"
                to="/settings/dictionaries/edit/thesaurusUnderscoreId2"
              >
                <Connect(Icon)
                  icon="pencil-alt"
                />
                 
                <span>
                  Edit
                </span>
              </Connect(I18NLink)>
                
              <button
                className="btn btn-danger btn-xs template-remove"
                onClick={[Function]}
                type="button"
              >
                <Connect(Icon)
                  icon="trash-alt"
                />
                 
                <span>
                  Delete
                </span>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="settings-footer"
  >
    <Connect(I18NLink)
      className="btn btn-success"
      to="/settings/dictionaries/new"
    >
      <Connect(Icon)
        icon="plus"
      />
      <span
        className="btn-label"
      >
        Add thesaurus
      </span>
    </Connect(I18NLink)>
  </div>
</div>
`;
